cmd/snap-update-ns/change.go

```diff
@@ -289,13 +289,31 @@ func (c *Change) lowLevelPerform(as *Assumptions) error {
 			// symlinks are handled in createInode directly, nothing to do here.
 		case "", "file":
 			flags, unparsed := osutil.MountOptsToCommonFlags(c.Entry.Options)
-			// Use Secure.BindMount for bind mounts
+			// Split the mount flags from the event propagation changes.
+			// Those have to be applied separately.
+			const propagationMask = syscall.MS_SHARED | syscall.MS_SLAVE | syscall.MS_PRIVATE | syscall.MS_UNBINDABLE
+			maskedFlagsRecursive := flags & syscall.MS_REC
+			maskedFlagsPropagation := flags & propagationMask
+			maskedFlagsNotPropagationNotRecursive := flags & ^(propagationMask | syscall.MS_REC)
+
 			if flags&syscall.MS_BIND == syscall.MS_BIND {
-				err = BindMount(c.Entry.Name, c.Entry.Dir, uint(flags))
+				// bind / rbind mount
+				flagsForMount := uintptr(maskedFlagsNotPropagationNotRecursive | maskedFlagsRecursive)
+				err = BindMount(c.Entry.Name, c.Entry.Dir, uint(flagsForMount))
+				logger.Debugf("mount %q %q %q %d %q (error: %v)", c.Entry.Name, c.Entry.Dir, c.Entry.Type, flagsForMount, strings.Join(unparsed, ","), err)
 			} else {
-				err = sysMount(c.Entry.Name, c.Entry.Dir, c.Entry.Type, uintptr(flags), strings.Join(unparsed, ","))
+				// normal mount, not bind / rbind, not propagation change
+				flagsForMount := uintptr(maskedFlagsNotPropagationNotRecursive)
+				err = sysMount(c.Entry.Name, c.Entry.Dir, c.Entry.Type, uintptr(flagsForMount), strings.Join(unparsed, ","))
+				logger.Debugf("mount %q %q %q %d %q (error: %v)", c.Entry.Name, c.Entry.Dir, c.Entry.Type, uintptr(flagsForMount), strings.Join(unparsed, ","), err)
```

Comment:
why have two almost-identical debugf calls? (and why are they only _almost_ identical?) I'd suggest making them actually identical, and moving them out of the if/else

Reply:
Yeah, perhaps you are right :) I'll look into this.

Reply:
Done now